File: R/dSepAdji.R
# If you have clean function-only versions of these files, you can source them here.
source("computePathMatrix.R")
source("computePathMatrix2.R")

dSepAdji <- function(AdjMat, i, condSet, PathMatrix = NULL, PathMatrix2 = NULL, spars = NULL)
    # Copyright (c) 2013 - 2014  Jonas Peters  [peters@stat.math.ethz.ch]
    # All rights reserved.  See the file COPYING for license terms.
    # This function looks for all j, such that i is d-sep to j given the condSet
    # The PathMatrix contains ancestor-relations.
    #
    # REACHABLE:
    #      means that there is a path from i that is not blocked by condSet
    # REACHABLEONNONCAUSALPATH = REACHABLEONNONDIRECTEDPATH (both notations exist):
    #      means that there is a path from i that is not blocked by condSet and that is NOT directed
    #
    # PathMatrix2 contains ancestor-relations when removing arrows condSet->
    # reachableOnNonCausalPath indicates whether in AdjMat there is a nonDirected path from i to j that is not blocked by condSet.
    # More details: For all combinations of nodes and edge direction (incoming or outgoing), we check, to which
    # neighbouring node-orientation pair an open path can be propagated further.
{
    p <- dim(AdjMat)[2]

    if (is.null(spars)) {
        spars <- (p > 99)
    }

    if (is.null(PathMatrix)) {
        PathMatrix <- computePathMatrix(AdjMat)
    }

    if (is.null(PathMatrix2)) {
        PathMatrix2 <- matrix(NA, p, p)
    }

    timeComputePM2 <- 0
    timeComputePM <- 0

    if(is.na(sum(PathMatrix2)))
    {
        ptm <- proc.time()
        PathMatrix2 <- computePathMatrix2(AdjMat, condSet, PathMatrix)
        timeComputePM2 <- timeComputePM2 + (proc.time() - ptm)[3]
    }

    if(length(condSet) == 0)
    {
        AncOfCondSet <- c()
    }

    if(length(condSet) == 1)
    {
        AncOfCondSet <- which(PathMatrix[, condSet] > 0)
    }

    if(length(condSet) > 1)
    {
        AncOfCondSet <- which(rowSums(PathMatrix[, condSet]) > 0)
    }

    reachabilityMatrix <- matrix(0, 2 * p, 2 * p)
    reachableOnNonCausalPathLater <- matrix(0, 2, 2)

    reachableNodes <- rep(0, 2 * p)
    reachableOnNonCausalPath <- rep(0, 2 * p)
    alreadyChecked <- rep(0, p)
    k <- 2
    toCheck <- c(0, 0)

    reachableCh <- which(AdjMat[i, ] == 1)
    if(length(reachableCh) > 0)
    {
        toCheck <- c(toCheck, reachableCh)
        reachableNodes[reachableCh] <- rep(1, length(reachableCh))
        AdjMat[i, reachableCh] <- rep(0, length(reachableCh))
    }

    reachablePa <- which(AdjMat[, i] == 1)
    if(length(reachablePa) > 0)
    {
        toCheck <- c(toCheck, reachablePa)
        reachableNodes[reachablePa + p] <- rep(1, length(reachablePa))
        reachableOnNonCausalPath[reachablePa + p] <- rep(1, length(reachablePa))
        AdjMat[reachablePa, i] <- rep(0, length(reachablePa))
    }

    while(k < length(toCheck))
    {
        k <- k + 1
        a1 <- toCheck[k]

        if(alreadyChecked[a1] == 0)
        {
            currentNode <- a1
            alreadyChecked[a1] <- 1

            Pa <- which(AdjMat[, currentNode] == 1)

            Pa1 <- setdiff(Pa, condSet)
            reachabilityMatrix[Pa1, currentNode] <- rep(1, length(Pa1))
            reachabilityMatrix[Pa1 + p, currentNode] <- rep(1, length(Pa1))

            if(sum(AncOfCondSet == currentNode) > 0)
            {
                reachabilityMatrix[currentNode, Pa + p] <- rep(1, length(Pa))

                if(PathMatrix2[i, currentNode] > 0)
                {
                    reachableOnNonCausalPathLater <- rbind(
                        reachableOnNonCausalPathLater,
                        cbind(rep(currentNode, length(Pa)), Pa)
                    )
                }

                newtoCheck <- Pa
                newtoCheck <- newtoCheck[which(alreadyChecked[newtoCheck] == 0)]
                toCheck <- c(toCheck, newtoCheck)
            }

            if(sum(condSet == currentNode) == 0)
            {
                reachabilityMatrix[currentNode + p, Pa + p] <- rep(1, length(Pa))
                newtoCheck <- Pa
                newtoCheck <- newtoCheck[which(alreadyChecked[newtoCheck] == 0)]
                toCheck <- c(toCheck, newtoCheck)
            }

            Ch <- which(AdjMat[currentNode, ] == 1)

            Ch1 <- setdiff(Ch, condSet)
            reachabilityMatrix[Ch1 + p, currentNode + p] <- rep(1, length(Ch1))

            Ch2 <- intersect(Ch, AncOfCondSet)
            reachabilityMatrix[Ch2, currentNode + p] <- rep(1, length(Ch2))
            Ch2b <- intersect(Ch2, which(PathMatrix2[i, ] > 0))
            reachableOnNonCausalPathLater <- rbind(
                reachableOnNonCausalPathLater,
                cbind(Ch2b, rep(currentNode, length(Ch2b)))
            )

            if(sum(condSet == currentNode) == 0)
            {
                reachabilityMatrix[currentNode, Ch] <- rep(1, length(Ch))
                reachabilityMatrix[currentNode + p, Ch] <- rep(1, length(Ch))
                newtoCheck <- Ch
                newtoCheck <- newtoCheck[which(alreadyChecked[newtoCheck] == 0)]
                toCheck <- c(toCheck, newtoCheck)
            }
        }
    }

    ptm <- proc.time()
    reachabilityMatrix <- computePathMatrix(reachabilityMatrix, spars = spars)
    timeComputePM <- timeComputePM + (proc.time() - ptm)[3]
    reachabilityMatrix <- as(reachabilityMatrix, "matrix")

    ttt2 <- which(reachableNodes == 1)
    if(length(ttt2) == 1)
    {
        tt2 <- which(reachabilityMatrix[ttt2, ] > 0)
    } else
    {
        tt2 <- which(colSums(reachabilityMatrix[ttt2, ]) > 0)
    }
    reachableNodes[tt2] <- rep(1, length(tt2))

    ttt <- which(reachableOnNonCausalPath == 1)
    if(length(ttt) == 1)
    {
        tt <- which(reachabilityMatrix[ttt, ] > 0)
    } else
    {
        tt <- which(colSums(reachabilityMatrix[ttt, ]) > 0)
    }
    reachableOnNonCausalPath[tt] <- rep(1, length(tt))

    if(dim(reachableOnNonCausalPathLater)[1] > 2)
    {
        for(kk in 3:(dim(reachableOnNonCausalPathLater)[1]))
        {
            ReachableThrough <- reachableOnNonCausalPathLater[kk, 1]
            newReachable <- reachableOnNonCausalPathLater[kk, 2]
            reachableOnNonCausalPath[newReachable + p] <- 1

            reachabilityMatrix[newReachable, ReachableThrough] <- 0
            reachabilityMatrix[newReachable, ReachableThrough + p] <- 0
            reachabilityMatrix[newReachable + p, ReachableThrough] <- 0
            reachabilityMatrix[newReachable + p, ReachableThrough + p] <- 0
        }

        ttt <- which(reachableOnNonCausalPath == 1)
        if(length(ttt) == 1)
        {
            tt <- which(reachabilityMatrix[ttt, ] > 0)
        } else
        {
            tt <- which(colSums(reachabilityMatrix[ttt, ]) > 0)
        }
        reachableOnNonCausalPath[tt] <- rep(1, length(tt))
    }

    result <- list()
    result$timeComputePM <- timeComputePM
    result$timeComputePM2 <- timeComputePM2
    result$reachableJ <- rowSums(cbind(reachableNodes[1:p], reachableNodes[(p + 1):(2 * p)])) > 0
    result$reachableOnNonCausalPath <- rowSums(cbind(reachableOnNonCausalPath[1:p], reachableOnNonCausalPath[(p + 1):(2 * p)])) > 0

    return(result)
}

#llm
read_testcase <- function(filepath)
{
    lines <- readLines(filepath, warn = FALSE)

    AdjMat <- list()
    condSet <- integer(0)
    i <- NULL
    mode <- NULL

    for(line in lines)
    {
        line <- trimws(line)

        if(line == "")
        {
            next
        }

        if(line == "Matrix:")
        {
            mode <- "matrix"
            next
        }

        if(line == "i:")
        {
            mode <- "i"
            next
        }

        if(line == "condSet:")
        {
            mode <- "condSet"
            next
        }

        if(mode == "matrix")
        {
            AdjMat[[length(AdjMat) + 1]] <- as.integer(strsplit(line, " +")[[1]])
        } else if(mode == "i")
        {
            i <- as.integer(line)
        } else if(mode == "condSet")
        {
            if(line != "empty")
            {
                condSet <- as.integer(strsplit(line, " +")[[1]])
            }
        }
    }

    AdjMat <- do.call(rbind, AdjMat)

    return(list(
        AdjMat = AdjMat,
        i = i,
        condSet = condSet
    ))
}


script_dir <- getwd()
project_dir <- dirname(script_dir)

testcase_dir <- file.path(project_dir, "tests", "dSepAdji", "Testcase")
output_dir <- file.path(project_dir, "tests", "dSepAdji", "R_outputs")

if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
}

output_file <- file.path(output_dir, "all_results.txt")
out <- file(output_file, open = "w")

for(i_case in 1:10000)
{
    filepath <- file.path(testcase_dir, paste0(i_case, ".txt"))
    values <- read_testcase(filepath)

    AdjMat <- values$AdjMat
    i <- values$i
    condSet <- values$condSet

    result <- dSepAdji(AdjMat, i, condSet)

    writeLines(paste0("Testcase ", i_case), out)

    writeLines("Matrix:", out)
    for(r in 1:nrow(AdjMat)) {
        writeLines(paste(AdjMat[r, ], collapse = " "), out)
    }

    writeLines("i:", out)
    writeLines(as.character(i), out)

    writeLines("condSet:", out)
    if(length(condSet) == 0) {
        writeLines("empty", out)
    } else {
        writeLines(paste(condSet, collapse = " "), out)
    }

    writeLines("reachableJ:", out)
    writeLines(paste(as.integer(result$reachableJ), collapse = " "), out)

    writeLines("reachableOnNonCausalPath:", out)
    writeLines(paste(as.integer(result$reachableOnNonCausalPath), collapse = " "), out)

    writeLines("timeComputePM:", out)
    writeLines(as.character(result$timeComputePM), out)

    writeLines("timeComputePM2:", out)
    writeLines(as.character(result$timeComputePM2), out)

    writeLines("", out)
}

close(out)

File: py_impl/dSepAdji.py
import os
import time
import numpy as np
from computePathMatrix import compute_path_matrix
from computePathMatrix2 import compute_path_matrix2


def d_sep_adji(adj_mat, i, cond_set, path_matrix=None, path_matrix2=None, spars=None):
    adj_mat = np.array(adj_mat, dtype=int).copy()
    p = adj_mat.shape[1]

    if spars is None:
        spars = p > 99

    # llm
    if path_matrix is None:
        path_matrix = compute_path_matrix(adj_mat, spars=spars)

    if path_matrix2 is None:
        path_matrix2 = np.full((p, p), np.nan)

    time_compute_pm2 = 0.0
    time_compute_pm = 0.0

    if np.isnan(path_matrix2).any():
        start = time.time()
        path_matrix2 = compute_path_matrix2(adj_mat.copy(), cond_set, path_matrix, spars=spars)
        time_compute_pm2 += time.time() - start

    if len(cond_set) == 0:
        anc_of_cond_set = np.array([], dtype=int)
    elif len(cond_set) == 1:
        anc_of_cond_set = np.where(path_matrix[:, cond_set[0] - 1] > 0)[0] + 1
    else:
        cols = [x - 1 for x in cond_set]
        anc_of_cond_set = np.where(np.sum(path_matrix[:, cols], axis=1) > 0)[0] + 1

    reachability_matrix = np.zeros((2 * p, 2 * p), dtype=int)
    reachable_on_noncausal_path_later = np.zeros((2, 2), dtype=int)

    reachable_nodes = np.zeros(2 * p, dtype=int)
    reachable_on_noncausal_path = np.zeros(2 * p, dtype=int)
    already_checked = np.zeros(p, dtype=int)

    k = 1
    to_check = [0, 0]

    i_idx = i - 1

    reachable_ch = np.where(adj_mat[i_idx, :] == 1)[0] + 1
    if len(reachable_ch) > 0:
        to_check.extend(reachable_ch.tolist())
        reachable_nodes[reachable_ch - 1] = 1
        adj_mat[i_idx, reachable_ch - 1] = 0

    reachable_pa = np.where(adj_mat[:, i_idx] == 1)[0] + 1
    if len(reachable_pa) > 0:
        to_check.extend(reachable_pa.tolist())
        reachable_nodes[reachable_pa - 1 + p] = 1
        reachable_on_noncausal_path[reachable_pa - 1 + p] = 1
        adj_mat[reachable_pa - 1, i_idx] = 0

    while k < len(to_check) - 1:
        k += 1
        a1 = to_check[k]

        if already_checked[a1 - 1] == 0:
            current_node = a1
            current_idx = current_node - 1
            already_checked[a1 - 1] = 1

            #parents of current node #llm
            pa = np.where(adj_mat[:, current_idx] == 1)[0] + 1

            pa1 = np.array([x for x in pa if x not in cond_set], dtype=int)
            if len(pa1) > 0:
                reachability_matrix[pa1 - 1, current_idx] = 1
                reachability_matrix[pa1 - 1 + p, current_idx] = 1

            if np.any(anc_of_cond_set == current_node):
                if len(pa) > 0:
                    reachability_matrix[current_idx, pa - 1 + p] = 1

                if path_matrix2[i_idx, current_idx] > 0 and len(pa) > 0:
                    add_rows = np.column_stack(
                        (np.full(len(pa), current_node, dtype=int), pa)
                    )
                    reachable_on_noncausal_path_later = np.vstack(
                        (reachable_on_noncausal_path_later, add_rows)
                    )

                new_to_check = [x for x in pa if already_checked[x - 1] == 0]
                to_check.extend(new_to_check)

            if current_node not in cond_set and len(pa) > 0:
                reachability_matrix[current_idx + p, pa - 1 + p] = 1
                new_to_check = [x for x in pa if already_checked[x - 1] == 0]
                to_check.extend(new_to_check)

            #children of current node
            ch = np.where(adj_mat[current_idx, :] == 1)[0] + 1

            ch1 = np.array([x for x in ch if x not in cond_set], dtype=int)
            if len(ch1) > 0:
                reachability_matrix[ch1 - 1 + p, current_idx + p] = 1

            ch2 = np.array([x for x in ch if x in anc_of_cond_set], dtype=int)
            if len(ch2) > 0:
                reachability_matrix[ch2 - 1, current_idx + p] = 1

            reachable_from_i = np.where(path_matrix2[i_idx, :] > 0)[0] + 1
            ch2b = np.array([x for x in ch2 if x in reachable_from_i], dtype=int)
            if len(ch2b) > 0:
                add_rows = np.column_stack(
                    (ch2b, np.full(len(ch2b), current_node, dtype=int))
                )
                reachable_on_noncausal_path_later = np.vstack(
                    (reachable_on_noncausal_path_later, add_rows)
                )

            if current_node not in cond_set and len(ch) > 0:
                reachability_matrix[current_idx, ch - 1] = 1
                reachability_matrix[current_idx + p, ch - 1] = 1
                new_to_check = [x for x in ch if already_checked[x - 1] == 0]
                to_check.extend(new_to_check)

    start = time.time()
    reachability_matrix = compute_path_matrix(reachability_matrix, spars=spars)
    time_compute_pm += time.time() - start
    reachability_matrix = np.array(reachability_matrix, dtype=int)

    ttt2 = np.where(reachable_nodes == 1)[0]
    if len(ttt2) == 1:
        tt2 = np.where(reachability_matrix[ttt2[0], :] > 0)[0]
    elif len(ttt2) > 1:
        tt2 = np.where(np.sum(reachability_matrix[ttt2, :], axis=0) > 0)[0]
    else:
        tt2 = np.array([], dtype=int)
    reachable_nodes[tt2] = 1

    ttt = np.where(reachable_on_noncausal_path == 1)[0]
    if len(ttt) == 1:
        tt = np.where(reachability_matrix[ttt[0], :] > 0)[0]
    elif len(ttt) > 1:
        tt = np.where(np.sum(reachability_matrix[ttt, :], axis=0) > 0)[0]
    else:
        tt = np.array([], dtype=int)
    reachable_on_noncausal_path[tt] = 1

    if reachable_on_noncausal_path_later.shape[0] > 2:
        for kk in range(2, reachable_on_noncausal_path_later.shape[0]):
            reachable_through = reachable_on_noncausal_path_later[kk, 0]
            new_reachable = reachable_on_noncausal_path_later[kk, 1]

            reachable_on_noncausal_path[new_reachable - 1 + p] = 1

            rt = reachable_through - 1
            nr = new_reachable - 1

            reachability_matrix[nr, rt] = 0
            reachability_matrix[nr, rt + p] = 0
            reachability_matrix[nr + p, rt] = 0
            reachability_matrix[nr + p, rt + p] = 0

        ttt = np.where(reachable_on_noncausal_path == 1)[0]
        if len(ttt) == 1:
            tt = np.where(reachability_matrix[ttt[0], :] > 0)[0]
        elif len(ttt) > 1:
            tt = np.where(np.sum(reachability_matrix[ttt, :], axis=0) > 0)[0]
        else:
            tt = np.array([], dtype=int)
        reachable_on_noncausal_path[tt] = 1

    reachable_j = (
        np.column_stack((reachable_nodes[:p], reachable_nodes[p:2 * p])).sum(axis=1) > 0
    )
    reachable_on_noncausal = (
        np.column_stack((reachable_on_noncausal_path[:p], reachable_on_noncausal_path[p:2 * p])).sum(axis=1) > 0
    )

    result = {
        "timeComputePM": time_compute_pm,
        "timeComputePM2": time_compute_pm2,
        "reachableJ": reachable_j.astype(int),
        "reachableOnNonCausalPath": reachable_on_noncausal.astype(int),
    }

    return result


def read_testcase(filepath):
    adj_mat = []
    cond_set = []
    i = None
    mode = None

    with open(filepath, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()

            if line == "":
                continue

            if line == "Matrix:":
                mode = "matrix"
                continue

            if line == "i:":
                mode = "i"
                continue

            if line == "condSet:":
                mode = "condSet"
                continue

            if mode == "matrix":
                adj_mat.append([int(x) for x in line.split()])
            elif mode == "i":
                i = int(line)
            elif mode == "condSet":
                if line != "empty":
                    cond_set = [int(x) for x in line.split()]

    return np.array(adj_mat, dtype=int), i, cond_set


#llm

script_dir = os.getcwd()
project_dir = os.path.dirname(script_dir)

testcase_dir = os.path.join(project_dir, "tests", "dSepAdji", "Testcase")
output_dir = os.path.join(project_dir, "tests", "dSepAdji", "Py_outputs")

if not os.path.exists(output_dir):
    os.mkdir(output_dir)

output_file = os.path.join(output_dir, "all_results.txt")

print("Reading testcases from:", testcase_dir)
print("Saving output to:", output_file)

with open(output_file, "w", encoding="utf-8") as f:
    for case_index in range(1, 10001):
        filepath = os.path.join(testcase_dir, f"{case_index}.txt")
        adj_mat, i, cond_set = read_testcase(filepath)

        result = d_sep_adji(adj_mat, i, cond_set)

        f.write(f"Testcase {case_index}\n")

        f.write("Matrix:\n")
        for row in adj_mat:
            f.write(" ".join(map(str, row)) + "\n")

        f.write("i:\n")
        f.write(str(i) + "\n")

        f.write("condSet:\n")
        if len(cond_set) == 0:
            f.write("empty\n")
        else:
            f.write(" ".join(map(str, cond_set)) + "\n")

        f.write("reachableJ:\n")
        f.write(" ".join(map(str, result["reachableJ"])) + "\n")

        f.write("reachableOnNonCausalPath:\n")
        f.write(" ".join(map(str, result["reachableOnNonCausalPath"])) + "\n")

        f.write("timeComputePM:\n")
        f.write(str(result["timeComputePM"]) + "\n")

        f.write("timeComputePM2:\n")
        f.write(str(result["timeComputePM2"]) + "\n")

        f.write("\n")

print("Done. Output saved to:", output_file)